pacman/operations/rig_algorithms/routing_tree.py

```diff
@@ -0,0 +1,142 @@
+"""An explicit representation of a routing tree in a machine.
+
+This representation of a route explicitly describes a tree-structure and the
+complete path taken by a route. This is used during place and route in
+preference to a set of RoutingTableEntry tuples since it is more easily
+verified and more accurately represents the problem at hand.
+
+Based on
+https://github.com/project-rig/rig/blob/master/rig/place_and_route/routing_tree.py
+"""
+from collections import deque
+
+
+class RoutingTree(object):
+    """Explicitly defines a multicast route through a SpiNNaker machine.
+
+    Each instance represents a single hop in a route and recursively refers to
+    following steps.
+
+    Attributes
+    ----------
+    chip : (x, y)
+        The chip the route is currently passing through.
+    children : list
+        A :py:class:`list` of the next steps in the route represented by a
+        (route, object) tuple.
+
+        .. note::
+
+            Up until Rig 1.5.1 this structure used :py:class:`set`\ s to store
+            children. This was changed to :py:class:`list`\ s since sets incur
+            a large memory overhead and in practice the set-like behaviour of
+            the list of children is not useful.
+
+        The route must be either :py:class:`~rig.routing_table.Routes` or
+        `None`. If :py:class:`~rig.routing_table.Routes` then this indicates
+        the next step in the route uses a particular route.
+
+        The object indicates the intended destination of this step in the
+        route. It may be one of:
+
+        * :py:class:`~.rig.place_and_route.routing_tree.RoutingTree`
+          representing the continuation of the routing tree after following a
+          given link. (Only used if the :py:class:`~rig.routing_table.Routes`
+          object is a link and not a core).
+        * A vertex (i.e. some other Python object) when the route terminates at
+          the supplied vertex. Note that the direction may be None and so
+          additional logic may be required to determine what core to target to
+          reach the vertex.
+
+    See Also
+    --------
+    rig.routing_table.routing_tree_to_tables:
+        May be used to convert :py:class:`.RoutingTree` objects into routing
+        tables suitable for loading onto a SpiNNaker machine.
+    """  # noqa W605
+
+    # A *lot* of instances of this data structure are created and so its memory
+    # consumption is a sensitive matter. The following optimisations have been
+    # made:
+    # * Using __slots__ hugely reduces the size of instances of this class
+    #   object
+    # * Storing the chip coordinate as two values (_chip_x and _chip_y) rather
+    #   than a tuple saves 56 bytes per instance.
+    __slots__ = ["_chip_x", "_chip_y", "_children"]
+
+    def __init__(self, chip):
+        self.chip = chip
+        self._children = []
+
+    @property
+    def chip(self):
```

Comment:
redundant brackets

Reply:
technicaly yes but they do highlight that the return is a tuple so I like them